### PR TITLE
[INTERNAL] Add platform specific cpu relax method for various platforms.

### PR DESCRIPTION
**Pull request by rrahn**
Fixes hopefully some platform dependent issues described here: #1861

### CODE REVIEW

**Comment by mr-c** [comment]
Via https://buildd.debian.org/status/package.php?p=seqan2&suite=experimental this fixes the build for  powerpc, ppc64el (and others, probably)

but fails on sparc64: https://buildd.debian.org/status/fetch.php?pkg=seqan2&arch=sparc64&ver=2.3.0~0.201610110~af07bc1-dfsg-1&stamp=1476963695

```
cd "/<<PKGBUILDDIR>>/obj-sparc64-linux-gnu/apps/bs_tools" && /usr/bin/c++   -DSEQAN_APP_VERSION=\"0.1.6\" -DSEQAN_DATE=\"\" -DSEQAN_HAS_EXECINFO=1 -DSEQAN_HAS_ZLIB=1 -DSEQAN_REVISION=\"tarball\" -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -I"/<<PKGBUILDDIR>>/include"  -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2   -W -Wall -pedantic -fstrict-aliasing -Wstrict-aliasing    -DSEQAN_GLOBAL_EXCEPTION_HANDLER=1 -o CMakeFiles/casbar.dir/casbar.cpp.o -c "/<<PKGBUILDDIR>>/apps/bs_tools/casbar.cpp"
In file included from /<<PKGBUILDDIR>>/include/seqan/parallel.h:78:0,
                 from /<<PKGBUILDDIR>>/include/seqan/modifier.h:48,
                 from /<<PKGBUILDDIR>>/include/seqan/align.h:56,
                 from /<<PKGBUILDDIR>>/apps/pair_align/lib/pair_align_lib.h:41,
                 from /<<PKGBUILDDIR>>/apps/pair_align/lib/pair_align_global.cpp:37:
/<<PKGBUILDDIR>>/include/seqan/parallel/parallel_lock.h:221:12: error: missing binary operator before token "("
 #if definde(__SUNPRO_C)
```

File: include/seqan/parallel/parallel_lock.h [hunk 1]
```diff
@@ -213,14 +213,26 @@ struct ScopedWriteLock<TLock, Serial>
...
-    asm volatile ("yield" ::: "memory");
-#elif defined(STDLIB_VS)
+#if defined(STDLIB_VS)  // Visual Studio - all platforms.
     YieldProcessor();
-#elif defined(__SSE2__)
-    _mm_pause();
+#elif defined(__arm__) || defined(__aarch64__)  // ARM.
+    __asm__ __volatile__ ("yield" ::: "memory");
+#elif defined(__sparc) || defined(__sparc__)  // SPARC
+#if definde(__SUNPRO_C)
```
Comment (mr-c) [inline include/seqan/parallel/parallel_lock.h]: `s/definde/defined/` :-)

**Comment by rrahn** [comment]
Fixed it, thanks!

**Comment by h-2** [comment]
@mr-c If you confirm that this solves all our FTBFS on the more obscure cpu-types, I will merge this! Thanks.

File: include/seqan/parallel/parallel_lock.h [hunk 2]
```diff
@@ -213,14 +213,26 @@ struct ScopedWriteLock<TLock, Serial>
...
+    __asm __volatile__ ("rd %%ccr, %%g0\n\t"
+                        "rd %%ccr, %%g0\n\t"
+                        "rd %%ccr, %%g0")
 #else
-    __asm__ __volatile__("rep; nop" : : );
+    __asm __volatile__ ("rd %ccr, %g0\n\t"
+                        "rd %ccr, %g0\n\t"
+                        "rd %ccr, %g0")
+#endif  // defined(__SUNPRO_C)
+#elif defined(__ppc__) || defined(__ppc64__) || defined(_ARCH_PPC)  || defined(_ARCH_PPC64)  // PowerPC
```
Comment (h-2) [inline include/seqan/parallel/parallel_lock.h]: Can we change this to:
```c++
#elif defined(__powerpc__) || defined(__ppc__) || defined(__PPC__) // PowerPC
```
first is generic, second is for macOS, third is for OpenBSD
Taken from:
http://nadeausoftware.com/articles/2012/02/c_c_tip_how_detect_processor_type_using_compiler_predefined_macros

File: include/seqan/parallel/parallel_lock.h [hunk 3]
```diff
@@ -213,14 +213,26 @@ struct ScopedWriteLock<TLock, Serial>
...
-#if defined(__arm__) || defined(_M_ARM)
-    asm volatile ("yield" ::: "memory");
-#elif defined(STDLIB_VS)
+#if defined(STDLIB_VS)  // Visual Studio - all platforms.
     YieldProcessor();
-#elif defined(__SSE2__)
-    _mm_pause();
+#elif defined(__arm__) || defined(__aarch64__)  // ARM.
+    __asm__ __volatile__ ("yield" ::: "memory");
+#elif defined(__sparc) || defined(__sparc__)  // SPARC
```
Comment (h-2) [inline include/seqan/parallel/parallel_lock.h]: the first macro is sufficient.

**Comment by mr-c** [comment]
@h-2 Still [waiting](https://buildd.debian.org/status/package.php?p=seqan2&suite=experimental) on armhf, armel, mips, mips64el, mipsel, s390x, alpha, hurd-i386, powerpcspe, and ppc64.

All good on amd64, arm64, i386, powerpc, ppc64el, hppa, and x32!

**Comment by mr-c** [comment]
Additional successes: s390x, ppc64

Hurd failed as expected. Package version 2.3.0~0.20161019~af07bc1+dfsg-2 has been submitted with https://github.com/seqan/seqan/pull/1890 applied (and the typo fix for sparc64)

Still waiting on armhf, armel, mips, mips64el, mipsel, alpha, powerpcspe, and ppc64.

**Comment by mr-c** [comment]
FYI,
[mips build failure](https://buildd.debian.org/status/fetch.php?pkg=seqan2&arch=mips&ver=2.3.0~0.201610110~af07bc1-dfsg-1&stamp=1477142269) (similar to this PR w/o 6689c8a)
```
/usr/bin/c++   -g  -fdebug-prefix-map=/«PKGBUILDDIR»=. -fPIE -fstack-protector-strong -Wformat -Werror=format-security -O0 -mxgot -Wdate-time -D_FORTIFY_SOURCE=2   -W -Wall -pedantic -fstrict-aliasing -Wstrict-aliasing -fopenmp   -fPIE -pie -Wl,-z,relro -Wl,-z,now CMakeFiles/s4_join.dir/join.cpp.o  -o ../../bin/s4_join -rdynamic -lrt -lpthread 
CMakeFiles/razers3.dir/razers.cpp.o: In function `void std::call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*), std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*>(std::once_flag&, void (std::__future_base::_State_baseV2::*&&)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*), std::__future_base::_State_baseV2*&&, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*&&, bool*&&)':
/usr/include/c++/6/mutex:609:(.text._ZSt9call_onceIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS0_12_Result_baseENS4_8_DeleterEEvEEPbEJPS1_S9_SA_EEvRSt9once_flagOT_DpOT0_[_ZSt9call_onceIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS0_12_Result_baseENS4_8_DeleterEEvEEPbEJPS1_S9_SA_EEvRSt9once_flagOT_DpOT0_]+0x14c): relocation truncated to fit: R_MIPS_TLS_GOTTPREL against `std::__once_callable@@GLIBCXX_3.4.11'
/usr/include/c++/6/mutex:610:(.text._ZSt9call_onceIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS0_12_Result_baseENS4_8_DeleterEEvEEPbEJPS1_S9_SA_EEvRSt9once_flagOT_DpOT0_[_ZSt9call_onceIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS0_12_Result_baseENS4_8_DeleterEEvEEPbEJPS1_S9_SA_EEvRSt9once_flagOT_DpOT0_]+0x160): relocation truncated to fit: R_MIPS_TLS_GOTTPREL against `std::__once_call@@GLIBCXX_3.4.11'
CMakeFiles/razers3.dir/razers.cpp.o: In function `void std::call_once<void (std::thread::*)(), std::thread*>(std::once_flag&, void (std::thread::*&&)(), std::thread*&&)':
/usr/include/c++/6/mutex:609:(.text._ZSt9call_onceIMSt6threadFvvEJPS0_EEvRSt9once_flagOT_DpOT0_[_ZSt9call_onceIMSt6threadFvvEJPS0_EEvRSt9once_flagOT_DpOT0_]+0xe8): relocation truncated to fit: R_MIPS_TLS_GOTTPREL against `std::__once_callable@@GLIBCXX_3.4.11'
/usr/include/c++/6/mutex:610:(.text._ZSt9call_onceIMSt6threadFvvEJPS0_EEvRSt9once_flagOT_DpOT0_[_ZSt9call_onceIMSt6threadFvvEJPS0_EEvRSt9once_flagOT_DpOT0_]+0xfc): relocation truncated to fit: R_MIPS_TLS_GOTTPREL against `std::__once_call@@GLIBCXX_3.4.11'
CMakeFiles/razers3.dir/razers.cpp.o: In function `void std::__once_call_impl<std::_Bind_simple<std::_Mem_fn<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*)> (std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*)> >()':
/usr/include/c++/6/mutex:587:(.text._ZSt16__once_call_implISt12_Bind_simpleIFSt7_Mem_fnIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS2_12_Result_baseENS6_8_DeleterEEvEEPbEEPS3_SB_SC_EEEvv[_ZSt16__once_call_implISt12_Bind_simpleIFSt7_Mem_fnIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS2_12_Result_baseENS6_8_DeleterEEvEEPbEEPS3_SB_SC_EEEvv]+0x28): relocation truncated to fit: R_MIPS_TLS_GOTTPREL against `std::__once_callable@@GLIBCXX_3.4.11'
CMakeFiles/razers3.dir/razers.cpp.o: In function `void std::__once_call_impl<std::_Bind_simple<std::_Mem_fn<void (std::thread::*)()> (std::thread*)> >()':
/usr/include/c++/6/mutex:587:(.text._ZSt16__once_call_implISt12_Bind_simpleIFSt7_Mem_fnIMSt6threadFvvEEPS2_EEEvv[_ZSt16__once_call_implISt12_Bind_simpleIFSt7_Mem_fnIMSt6threadFvvEEPS2_EEEvv]+0x28): relocation truncated to fit: R_MIPS_TLS_GOTTPREL against `std::__once_callable@@GLIBCXX_3.4.11'
CMakeFiles/razers3.dir/razers.cpp.o:./obj-mips-linux-gnu/apps/razers3/./include/seqan/stream/buffered_stream.h:73:(.text._ZN5seqan14BufferedStreamISiNS_3TagINS_6Input_EEEED1Ev[_ZN5seqan14BufferedStreamISiNS_3TagINS_6Input_EEEED1Ev]+0x114): relocation truncated to fit: R_MIPS_GOT16 against `.text._ZN5seqan14BufferedStreamISiNS_3TagINS_6Input_EEEED1Ev'
CMakeFiles/razers3.dir/razers.cpp.o:./obj-mips-linux-gnu/apps/razers3/./include/seqan/stream/buffered_stream.h:73:(.text._ZN5seqan14BufferedStreamISiNS_3TagINS_6Input_EEEED0Ev[_ZN5seqan14BufferedStreamISiNS_3TagINS_6Input_EEEED0Ev]+0x98): relocation truncated to fit: R_MIPS_GOT16 against `.text._ZN5seqan14BufferedStreamISiNS_3TagINS_6Input_EEEED0Ev'
CMakeFiles/razers3.dir/razers.cpp.o:./obj-mips-linux-gnu/apps/razers3/./include/seqan/stream/virtual_stream.h:297:(.text._ZN5seqan13VirtualStreamIcNS_3TagINS_6Input_EEESt11char_traitsIcEED1Ev[_ZN5seqan13VirtualStreamIcNS_3TagINS_6Input_EEESt11char_traitsIcEED1Ev]+0x15c): relocation truncated to fit: R_MIPS_GOT16 against `.text._ZN5seqan13VirtualStreamIcNS_3TagINS_6Input_EEESt11char_traitsIcEED1Ev'
CMakeFiles/razers3.dir/razers.cpp.o:./obj-mips-linux-gnu/apps/razers3/./include/seqan/stream/virtual_stream.h:297:(.text._ZN5seqan13VirtualStreamIcNS_3TagINS_6Input_EEESt11char_traitsIcEED0Ev[_ZN5seqan13VirtualStreamIcNS_3TagINS_6Input_EEESt11char_traitsIcEED0Ev]+0x98): relocation truncated to fit: R_MIPS_GOT16 against `.text._ZN5seqan13VirtualStreamIcNS_3TagINS_6Input_EEESt11char_traitsIcEED0Ev'
collect2: error: ld returned 1 exit status
apps/razers3/CMakeFiles/razers3.dir/build.make:125: recipe for target 'bin/razers3' failed
```

**Comment by h-2** [comment]
Indeed the mips failure looks like something different. A quick search revealed that either of these CXX_FLAGS might "fix" it:
`-xgot` (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=274090)
`-mxgot` (https://lists.exim.org/lurker/message/20120724.163838.8fcfc213.en.html)
`-G4`, `-G2`, or `-G1` (https://www.linux-mips.org/archives/linux-mips/2003-05/msg00231.html)
Note that I have no clue about MIPS and what these flags actually do.

**Pull request merged**